### PR TITLE
Add Llama test for fsdp

### PR DESCRIPTION
**Pull request by hlahkar**
Add Llama test for FSDP

### CODE REVIEW

**Comment by hlahkar** [comment]
@regisss please review this test case. Our end goal is to test llama2; but we need to pass a auth token for that. We would like to know whether we can pass a token to the test case; in that case we can change the test from Llama to Llama2.

File: examples/language-modeling/README.md [hunk 1]
```diff
@@ -559,6 +559,47 @@ python3 ../gaudi_spawn.py --use_deepspeed  --world_size 8  run_lora_clm.py \
...
 ```
 
+- Multi-card finetuning of Llama2-70B with FSDP and LoRA:
+
+
+```bash
+PT_HPU_LAZY_MODE=0 LOWER_LIST=ops_bf16.txt \
+python3 ../gaudi_spawn.py --use_mpi  --world_size 8  run_lora_clm.py \
+  --model_name_or_path meta-llama/Llama-2-70b-hf \
+  --deepspeed llama2_ds_zero3_config.json \
```
Comment (greg-serochi) [inline examples/language-modeling/README.md:569]: why is the deepspeed json file being called out here?

File: examples/language-modeling/README.md [hunk 2]
```diff
@@ -559,6 +559,47 @@ python3 ../gaudi_spawn.py --use_deepspeed  --world_size 8  run_lora_clm.py \
...
+  --pipelining_fwd_bwd \
+  --throughput_warmup_steps 3 \
+  --lora_rank 4 \
+  --lora_target_modules "q_proj" "v_proj" "k_proj" "o_proj" \
+  --validation_split_percentage 4 \
+  --fsdp_config fsdp_config.json \
+  --fsdp "auto_wrap" \
+  --torch_compile_backend hpu_backend \
+  --torch_compile
+```
```
Comment (vivekgoe) [inline examples/language-modeling/README.md:601]: FusedRoPE = false is missing.

**Review by vivekgoe — CHANGES_REQUESTED** [review]

File: examples/language-modeling/README.md (hunk 2 above)
Comment (hlahkar) [reply examples/language-modeling/README.md:601]: Added

**Comment by regisss** [comment]
> @regisss please review this test case. Our end goal is to test llama2; but we need to pass a auth token for that. We would like to know whether we can pass a token to the test case; in that case we can change the test from Llama to Llama2.

Yes that should be feasible. Take a look at how it is done in https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py.
Basically your tests functions needs a `token` argument and you pass the token to pytest running:
```
pytest tests/test_fsdp_examples.py --token my_token
```
Let me know if that works.

**Review by regisss — APPROVED** [review]
LGTM!

I pushed a new commit to add the token arg in the CI command.

**Comment by regisss** [comment]
@hlahkar When I run
```
pytest tests/test_fsdp_examples.py -v -s -k "llama" --token my_token
```
the test fails because the train loss is nan. Can you run it and see if you get the same please?

**Comment by hlahkar** [comment]
@regisss on latest optimum-habana (after upgrade to transformer 4.38) this test is failing; prior to that this test was passing. Also this test need to be tested on Gaudi2 also is advisable to test this on 1.15 synapse release

**Comment by regisss** [comment]
> @regisss on latest optimum-habana (after upgrade to transformer 4.38) this test is failing; prior to that this test was passing. Also this test need to be tested on Gaudi2 also is advisable to test this on 1.15 synapse release

On Synapse 1.15 does it pass?

**Comment by hlahkar** [comment]
> > @regisss on latest optimum-habana (after upgrade to transformer 4.38) this test is failing; prior to that this test was passing. Also this test need to be tested on Gaudi2 also is advisable to test this on 1.15 synapse release
> 
> On Synapse 1.15 does it pass?

@regisss On Synapse 1.15 it passes with optimum-habana 1.10.4 + https://github.com/huggingface/optimum-habana/pull/753/files

But it fails with the latest optimum habana because of upgrade to transformers 4.38

**Comment by regisss** [comment]
> > > @regisss on latest optimum-habana (after upgrade to transformer 4.38) this test is failing; prior to that this test was passing. Also this test need to be tested on Gaudi2 also is advisable to test this on 1.15 synapse release
> > 
> > 
> > On Synapse 1.15 does it pass?
> 
> @regisss On Synapse 1.15 it passes with optimum-habana 1.10.4 + https://github.com/huggingface/optimum-habana/pull/753/files
> 
> But it fails with the latest optimum habana because of upgrade to transformers 4.38

Okay. With Synapse 1.14 it fails even before the upgrade to Transformers 4.38.
Can you update your main branch and rebase this PR on it so that we can find a fix with the latest codebase?

**Comment by regisss** [comment]
It works in fp32. There must be an operation between a bf16 and a fp32 somewhere that leads to nan.

**Comment by hlahkar** [comment]
@regisss I tested the test case on top of latest optimum-habana + https://github.com/huggingface/optimum-habana/pull/812; the test is passing

**Pull request merged**